### ISSUE TITLE
zsh extremely slow to start without internet connectivity

### ISSUE THREAD

**Issue by protist** [zimfw/zimfw#514]
- [x] I've checked the [existing issues](../issues?q=is%3Aissue) and I'm not duplicating a report.
- [ ] I'm using the [latest version](https://github.com/zimfw/zimfw/releases/latest) of `zimfw`. ---> **NO** I'm running the latest commits; see below.
- [x] I've checked the [Changelog](https://github.com/zimfw/zimfw/blob/master/CHANGELOG.md) and I'm not being affected by documented changes.
- [x] I've checked the [📢 Announcements](https://github.com/zimfw/zimfw/discussions/426?sort=new) and I'm not being affected by announced changes.
- [x] I was able to reproduce the issue with a clean installation of Zim.
- [x] I've pasted the output of `zimfw info` below.

**Describe the bug**
This has taken me a few hours to debug, but it might not be present in the latest version. Apologies if this is noise, but just in case it is still relevant, I wanted to submit this issue.

I had been experiencing this issue in a recent version of zim db96076, however, after updating to the latest version 0566b9b (two commits later), I am no longer experiencing this issue. This issue might have been fixed in these two commits, although from a skim, I don't think so. Alternatively, it is possible that the issue was *because* I was not running the latest version, since it relates to version checks? If so, then this issue is still relevant, and might re-emerge with new commit(s). If not, then apologies again for the noise!

I will describe the original issue here, experienced with the older version of zim. If I disconnect my network, zsh would take over two minutes to start up. `zsh -x` showed that it was stalling in the version check, which I presume needs internet access.

```
$ zsh -x
+/etc/zsh/zshrc:6> export ZIM_HOME=/usr/lib/zim
+/etc/zsh/zshrc:15> source /usr/lib/zim/zimfw.zsh init -q
+/usr/lib/zim/zimfw.zsh:27> autoload -Uz is-at-least
+/usr/lib/zim/zimfw.zsh:27> is-at-least 5.2
```

I tried commenting out the [relevant lines](https://github.com/zimfw/zimfw/blob/master/zimfw.zsh#L27-L29). It now progressed past this, but stalled elsewhere.

```
$ zsh -x
+/etc/zsh/zshrc:6> export ZIM_HOME=/usr/lib/zim
+/etc/zsh/zshrc:15> source /usr/lib/zim/zimfw.zsh init -q
+/usr/lib/zim/zimfw.zsh:31> autoload -Uz zargs
+/usr/lib/zim/zimfw.zsh:34> ((  ! 1  ))
+/usr/lib/zim/zimfw.zsh:939> [[ /etc/zsh/zshrc:15 == zmodule:* ]]
+/usr/lib/zim/zimfw.zsh:942> zimfw init -q
```

I tried searching through the source for this part, but got confused here.

FWIW I've installed system-wise through the Arch Linux [AUR package](https://aur.archlinux.org/packages/zsh-zim-git/). This (probably) meant I couldn't do the troubleshooting steps below unfortunately. However, hopefully you can just replicate it on your system.

**Steps to reproduce**  
The fist 4 steps restart the shell with a clean installation of Zim in a temporary directory.
Use `exec zsh` when restarting the terminal or restarting the shell is needed.
1. `cd ${$(mktemp -d):A}`
2. `ZDOTDIR=${PWD} HOME=${PWD} ZIM_HOME=${PWD}/.zim exec zsh`
3. `curl -fsSL https://raw.githubusercontent.com/zimfw/install/master/install.zsh | zsh`
```
% curl -fsSL https://raw.githubusercontent.com/zimfw/install/master/install.zsh | zsh
) Using Zsh version 5.9
) Your ZIM_HOME is customized to /usr/lib/zim
x /usr/lib/zim already exists. Please set ZIM_HOME to the path where you want to install Zim.
```

5. I was unable to continue with `exec zsh`

**Current behavior**  
Without internet access, zsh takes over two minutes to start. Presumably this will also be an issue with slow connections.

**Expected behavior**  
zsh to start almost instantaneously

**zimfw info**  
Recent version db96076, with bug.
```
# In a terminal, run zimfw info, and paste the output here.
$ zimfw info
_zimfw_check_version:6: permission denied: /usr/lib/zim/.latest_version
zimfw version:        1.12.0-SNAPSHOT (built at 2023-04-01 01:32:16 UTC, previous commit is 5db1155)
LANG:                 en_AU.UTF-8
OSTYPE:               linux-gnu
TERM:                 xterm-256color
TERM_PROGRAM:         
TERM_PROGRAM_VERSION: 
ZIM_HOME:             /usr/lib/zim
ZSH_VERSION:          5.9
```
BTW I have no idea why it references 5db1155, since I can't find this commit.

Current version 0566b9b, with no bug.
```
Latest zimfw version is 1.11.3. You're using version 1.12.0-SNAPSHOT. Run zimfw upgrade to upgrade.
zimfw version:        1.12.0-SNAPSHOT (built at 2023-06-15 14:32:20 UTC, previous commit is f9648e0)
LANG:                 en_AU.UTF-8
OSTYPE:               linux-gnu
TERM:                 xterm-256color
TERM_PROGRAM:         
TERM_PROGRAM_VERSION: 
ZIM_HOME:             /usr/lib/zim
ZSH_VERSION:          5.9
```

**Comment by ericbn** [comment]
Hi @protist. Thank you for the detailed issue description.

Let me investigate, but I can already say I've changed the automatic version check to not be asynchronous (not run in the backgroung) anymore in version [`a34b8da`](https://github.com/zimfw/zimfw/commit/a34b8dab64214216ffb27c00474d7c8314b1c98e), before `db96076`, where you mention the bug is present:

```
* 0566b9b Use _zimfw_print (HEAD -> master, origin/master, origin/HEAD)
* ...
* db96076 fixup! Don't use a new shell to run tool anymore
* ...
* a34b8da Add version-check action
```

The version check uses `git` to make a request to GitHub and now I see it was a bad idea to make the automatic version check synchronous, because it will make `zimfw` wait until the connection times out in case of a connection issue.

Maybe there's something in `db96076` that makes the problem worse. And not sure what makes it better again in `0566b9b`. I'll check and most probably make the automatic version check asynchronous again anyway.

**Comment by ericbn** [comment]
I believe I've fixed in https://github.com/zimfw/zimfw/commit/49386cad6def77fb335278b7ee641212c41d545b. Let me know if you still face issues.

And thanks for reporting this before a version containing the issue were released.

**Comment by protist** [comment]
No worries, and thanks for the amazingly quick fix @ericbn. I'm not entirely sure how to test this, but I disconnected from the internet then opened up a new terminal, and there was no slowdown (while running the previous 0566b9b8a35d5175820113bc91bf0c3465ca9479). I also upgraded and tested the latest version, and there is no slowdown there either. Presumably all is good, but I'll report back if I see anything weird. Thanks again!
